pyteal/ast/multi.py

```diff
@@ -0,0 +1,85 @@
+from typing import Callable, List, Union, TYPE_CHECKING
+
+from ..types import TealType
+from ..ir import TealOp, Op, TealBlock
+from .expr import Expr
+from .leafexpr import LeafExpr
+from .scratch import ScratchSlot
+
+if TYPE_CHECKING:
+    from ..compiler import CompileOptions
+
+
+class MultiValue(LeafExpr):
+    """Represents an operation that returns more than one value"""
+
+    def __init__(
+        self,
+        op: Op,
+        types: List[TealType],
+        *,
+        immediate_args: List[Union[int, str]] = None,
+        args: List[Expr] = None
+    ):
+        """Create a new MultiValue.
+
+        Args:
+            op: The operation that returns values.
+            types: The types of the returned values.
+            immediate_args (optional): Immediate arguments for the op. Defaults to None.
+            args (optional): Stack arguments for the op. Defaults to None.
+        """
+        super().__init__()
+        self.op = op
+        self.types = types
+        self.immediate_args = immediate_args if immediate_args is not None else []
+        self.args = args if args is not None else []
+
+        self.output_slots = [ScratchSlot() for _ in self.types]
+        self.reducer_set = False
+
+    def outputReducer(self, reducer: Callable[..., Expr]) -> Expr:
```

Comment:
@algoidurovic For my understanding:  What is the tradeoff between exposing _reducer_ as a constructor argument vs a separate method (like the PR)?

I'm trying to gauge if/how requiring the _reducer_ at construction complicates the picture.  Probably because my understanding is lacking, I am wondering _when_ the caller benefits from the separate method to customize the _reducer_.

Reply:
Aside:  I see the test cases reuse a `MultiValue` instance and change the _reducer_.  It's unclear to me if we'd expect regular re-use in production applications.

Reply:
1. I think it makes sense stylistically to "chain" a transformation function/reducer rather than pass it to the constructor. It meshes better with PyTeal code style imo.
2. The user doesn't usually construct the MultiValue/MaybeValue directly. A higher level function (for example, AccountParams.balance(..)) encapsulates the construction of the MaybeValue so the user couldn't pass in a reducer.

I wouldn't expect the reducer to be redefined in production apps very often but that option is available. It could be useful if you want to apply more than one transformation to the outputs in a chain or separately. As you pointed out, it's nice for testing as well.

Reply:
@algoidurovic Got it - thanks for the context.  All set on my end - resolving.